filecoin-proofs/src/bin/paramfetch.rs

```diff
@@ -81,77 +120,162 @@ fn fetch(matches: &ArgMatches) -> Result<()> {
         )));
     }
 
-    let parameter_map = get_parameter_map(&json_path)?;
-    let all_parameter_ids = get_mapped_parameter_ids(&parameter_map)?;
+    let manifest = read_parameter_map_from_disk(&json_path)?;
+    let mut filenames = get_filenames_from_parameter_map(&manifest)?;
+
+    println!("{} files in manifest...", filenames.len());
+    println!();
+
+    // if user has specified sector sizes for which they wish to download Groth
+    // parameters, trim non-matching Groth parameter filenames from the list
+    if matches.is_present("params-for-sector-sizes") {
+        let whitelisted_sector_sizes: Vec<u64> =
+            values_t!(matches.values_of("params-for-sector-sizes"), u64)?;
+
+        let whitelisted_sector_sizes: HashSet<u64> =
+            HashSet::from_iter(whitelisted_sector_sizes.iter().cloned());
+
+        // always download all verifying keys - but conditionally skip Groth
+        // parameters for sector sizes the user doesn't care about
+        filenames = filenames
+            .into_iter()
+            .filter(|id| {
+                !has_extension(id, GROTH_PARAMETER_EXT) || {
+                    manifest
+                        .get(id)
+                        .map(|p| p.sector_size)
+                        .map(|n| whitelisted_sector_sizes.contains(&n))
+                        .unwrap_or(false)
+                }
+            })
+            .collect_vec();
+    }
 
-    println!("checking {} parameter files...", all_parameter_ids.len());
+    println!("{} files to check for (re)download...", filenames.len());
     println!();
 
-    let mut parameter_ids = get_pending_parameter_ids(&parameter_map, all_parameter_ids.clone())?;
+    // ensure filename corresponds to asset on disk and that its checksum
+    // matches that which is specified in the manifest
+    filenames = get_filenames_requiring_download(&manifest, filenames)?;
 
-    if !matches.is_present("all") && !parameter_ids.is_empty() {
-        parameter_ids = choose_from(parameter_ids)?;
+    // don't prompt the user to download files if they've used certain flags
+    if !matches.is_present("params-for-sector-sizes")
+        && !matches.is_present("all")
+        && !filenames.is_empty()
+    {
+        filenames = choose_from(&filenames, |filename| {
+            manifest.get(filename).map(|x| x.sector_size)
+        })?;
         println!();
     }
 
     loop {
-        println!("{} parameter files to fetch...", parameter_ids.len());
+        println!("{} files to fetch...", filenames.len());
         println!();
 
-        for parameter_id in &parameter_ids {
-            println!("fetching: {}", parameter_id);
-            print!("downloading parameter file... ");
+        for filename in &filenames {
+            println!("fetching: {}", filename);
+            print!("downloading file... ");
             io::stdout().flush().unwrap();
```

Comment:
why is this flushing stdout manually?

Reply:
No idea. It's left over code from the original impl.